### PR TITLE
Allow applying normalization to GNN embeddings for link prediction.

### LINKED ISSUE #632: Apply normalization to GNN embeddings for link prediction tasks.
In some cases, link prediction is used as a unsupervised method to pre-train a GNN model and the generated node embeddings are required to be normalized, for example L2 norm. GraphStorm should support this capability.

### PR DESCRIPTION
**Pull request by classicsong**
Resolve #632 

Users can use `--lp-embed-normalizer l2_norm` to enable normalizing node embeddings with l2 norm during model training and inference. By default, it is set to None. (No normalization is applied).

Using l2_norm, the training can be more stable, especially with contrastive loss.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/model/graph_utils.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
+"""
```
Comment (zheng-da) [inline python/graphstorm/model/graph_utils.py:1]: do you make any change in this file? or you just moved the code

File: python/graphstorm/model/utils.py [hunk 2]
```diff
@@ -1686,3 +1688,45 @@ def append_to_dict(from_dict, to_dict):
...
+            if is_train is True:
+                # When it is training, we expect the size of emb is small
+                return F.normalize(emb)
+
+            rank = get_rank()
+            world_size = get_world_size()
+            if isinstance(emb, (dgl.distributed.DistTensor, LazyDistTensor)):
+                start, end = get_data_range(rank, world_size, len(emb))
+            else:
+                start, end = 0, len(emb)
```
Comment (zheng-da) [inline python/graphstorm/model/utils.py:1719]: when this happens? if emb is a normal tensor, why not call F.normalize directly?

File: python/graphstorm/model/graph_utils.py (hunk 1 above)
Comment (classicsong) [reply python/graphstorm/model/graph_utils.py:1]: The changes are L111-115 and L200-L204

File: python/graphstorm/model/utils.py [hunk 3]
```diff
@@ -1686,3 +1688,45 @@ def append_to_dict(from_dict, to_dict):
...
+            Node embedding normalization method.
+        is_train: bool
+            Whether it is called during training.
+    """
+    if norm_method is None or norm_method == "":
+        def norm(emb):
+            return emb
+        norm_func = norm
+    elif norm_method == GRAPHSTORM_LP_EMB_L2_NORMALIZATION:
+        def l2_norm(emb):
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/model/utils.py:1709]: OK

File: python/graphstorm/model/utils.py (hunk 2 above)
Comment (classicsong) [reply python/graphstorm/model/utils.py:1719]: In minibatch inference, it will be a normal tensor.
The tensor may be super large, call F.normalize may cause some problems.

File: python/graphstorm/model/graph_utils.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                                                model.gnn_encoder,
+                                                get_input_embeds,
+                                                batch_size, fanout,
+                                                edge_mask=edge_mask,
+                                                target_ntypes=infer_ntypes,
+                                                task_tracker=task_tracker)
+
+    if isinstance(model, GSgnnLinkPredictionModelInterface):
+        # If it is link prediction, try to normalize the node embeddings
+        embeddings = model.normalize_node_embs(embeddings)
```
Comment (zheng-da) [inline python/graphstorm/model/graph_utils.py:113]: should this be in the predict function?

File: python/graphstorm/model/graph_utils.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            return model.node_input_encoder(feats, input_nodes)
+
+        embeddings = model.gnn_encoder.dist_inference(data.g, get_input_embeds,
+                                                    batch_size, fanout, edge_mask=edge_mask,
+                                                    task_tracker=task_tracker)
+
+    if isinstance(model, GSgnnLinkPredictionModelInterface):
+        # If it is link prediction, try to normalize the node embeddings
+        # Called when model.eval()
+        embeddings = model.normalize_node_embs(embeddings)
```
Comment (zheng-da) [inline python/graphstorm/model/graph_utils.py:203]: can we merge it into the model? it'll make the code more modularized. also memory efficient.

Comment (classicsong) [reply python/graphstorm/model/graph_utils.py:203]: No. 
It is very hard to do so, which may need refactoring the entire inference code especially full_graph_inference.

File: python/graphstorm/model/gnn.py [hunk 6]
```diff
@@ -874,6 +895,7 @@ def do_mini_batch_inference(model, data, batch_size=1024,
                                                    task_tracker=task_tracker,
                                                    feat_field=data.node_feat_field,
                                                    target_ntypes=infer_ntypes)
+        model.eval()
```
Comment (zheng-da) [inline python/graphstorm/model/gnn.py:898]: why do you call eval here?

File: python/graphstorm/model/gnn.py [hunk 7]
```diff
@@ -914,6 +936,8 @@ def get_input_embeds(input_nodes):
                                                 edge_mask=edge_mask,
                                                 target_ntypes=infer_ntypes,
                                                 task_tracker=task_tracker)
+        # Called when model.eval()
+        embeddings = model.normalize_node_embs(embeddings)
```
Comment (zheng-da) [inline python/graphstorm/model/gnn.py:940]: you only call this function in this branch?

Comment (classicsong) [reply python/graphstorm/model/gnn.py:940]: We should call it for all branches.

Comment (classicsong) [reply python/graphstorm/model/gnn.py:940]: udpated

File: python/graphstorm/model/gnn.py (hunk 6 above)
Comment (classicsong) [reply python/graphstorm/model/gnn.py:898]: embeddings = model.normalize_node_embs(embeddings) should be called with model.eval()

File: python/graphstorm/model/lp_gnn.py [hunk 8]
```diff
@@ -94,6 +109,9 @@ def forward(self, blocks, pos_graph,
             # GNN message passing
             encode_embs = self.compute_embed_step(blocks, node_feats, input_nodes)
 
+        # Call emb normalization.
+        self.normalize_node_embs(encode_embs)
```
Comment (zheng-da) [inline python/graphstorm/model/lp_gnn.py:113]: what is this? why do you not get the returned embeddings?

Comment (classicsong) [reply python/graphstorm/model/lp_gnn.py]: Fixed

**Review by zheng-da — APPROVED** [review]
